Arm/Memory/AddressNormalization.lean

```diff
@@ -0,0 +1,267 @@
+-/
+def getBitVecOfNatValue? (e : Expr) : (Option (Expr × Expr)) :=
+  match_expr e with
+  | BitVec.ofNat nExpr vExpr => some (nExpr, vExpr)
+  | _ => none
+
+/-- Try to build a proof for `ty` by reduction to `bv_omega`. -/
+@[inline] def proveByBvOmega (ty : Expr) : SimpM Step := do
+  let proof : Expr ← mkFreshExprMVar ty
+  let g := proof.mvarId!
+  let some g ← g.falseOrByContra
+    | return .continue
+  try
+    g.withContext (do Lean.Elab.Tactic.Omega.omega (← getLocalHyps).toList g {})
+  catch _ =>
+    return .continue
+  return .done { expr := ty, proof? := proof }
+
+/-- Try to build a proof for `ty` by reduction to `omega`. -/
+@[inline] def proveByOmega (ty : Expr) : SimpM Step := do
+  let proof : Expr ← mkFreshExprMVar ty
+  let g := proof.mvarId!
+  let some g ← g.falseOrByContra
+    | return .continue
+  try
+    g.withContext (do Lean.Elab.Tactic.Omega.omega (← getLocalHyps).toList g {})
+  catch _ =>
+    return .continue
+  return .done { expr := ty, proof? := proof }
+
+def processingEmoji : String := "⚙️"
+
+-- x % n = x if x < n
+@[inline] def reduceModOfLt (x : Expr) (n : Expr) : SimpM Step := do
+  trace[Tactic.address_normalization] "{processingEmoji} reduceModOfLt '{x} % {n}'"
+  let ltTy := mkLTNat x n
+  let Step.done { expr := _, proof? := some p} ← proveByOmega ltTy
+    | return .continue
+  let eqProof ← mkAppM ``Nat.mod_eq_of_lt #[p]
+  trace[Tactic.address_normalization] "{checkEmoji} reduceModOfLt '{x} % {n}'"
+  return .done { expr := x, proof? := eqProof : Result }
+
+-- x % n = x - n if x >= n and x - n < n
+@[inline] def reduceModSub (x : Expr) (n : Expr) : SimpM Step := do
+  trace[Tactic.address_normalization] "{processingEmoji} reduceModSub '{x} % {n}'"
+  let geTy := mkGENat x n
+  let Step.done { expr := _, proof? := some geProof} ← proveByOmega geTy
+    | return .continue
+  let subTy := mkSubNat x n
+  let ltTy := mkLTNat subTy n
+  let Step.done { expr := _, proof? := some ltProof} ← proveByOmega ltTy
+    | return .continue
+  let eqProof ← mkAppM ``Nat.mod_eq_sub #[geProof, ltProof]
+  trace[Tactic.address_normalization] "{checkEmoji} reduceModSub '{x} % {n}'"
+  return .done { expr := subTy, proof? := eqProof : Result }
+
+@[inline, bv_toNat] def reduceMod (e : Expr) : SimpM Step := do
+  match_expr e with
+  | HMod.hMod xTy nTy outTy _inst x n =>
+    let natTy := mkConst ``Nat
+    if (xTy != natTy) || (nTy != natTy) || (outTy != natTy) then
+      return .continue
+    if let .done res ← reduceModOfLt x n then
+      return .done res
+    if let .done res ← reduceModSub x n then
+      return .done res
+    return .continue
+  | _ => do
+     return .continue
+
+simproc↑ [address_normalization] reduce_mod_omega (_ % _) := fun e => reduceMod e
+
+/-- In a commutative binary operation, move a bitvector constant to the left. -/
+@[inline, bv_toNat] def moveBinConstLeft (declName : Name) -- operator to constant fold, such as `HAdd.hAdd`.
```

Comment:
It's not just moving, it's also constant folding, right?

Let's call it `canonicalizeBinConst`, and explicitly mention in the docstring that this:
- will constant fold if both operands are constant, or
- move the constant left, if there is exactly one

Reply:
Done!